stream/stream_worker.go

```diff
@@ -19,116 +22,45 @@ type GripStream interface {
 // forwards SSEEvents from the embedded SDKs to and forwards them on to clients
 type StreamWorker struct {
-	log    log.Logger
-	gpc    GripStream
-	stream Stream
-	topics []string
+	log log.Logger
+	gpc GripStream
 }
 
 // NewStreamWorker creates a StreamWorker
-func NewStreamWorker(l log.Logger, gpc GripStream, stream Stream, topics ...string) StreamWorker {
+func NewStreamWorker(l log.Logger, gpc GripStream) StreamWorker {
 	l = l.With("component", "StreamWorker")
 	return StreamWorker{
-		log:    l,
-		gpc:    gpc,
-		stream: stream,
-		topics: topics,
+		log: l,
+		gpc: gpc,
 	}
 }
 
-// Run starts the stream workers process of listening for events from the passed
-// Stream and forwarding them on to the passed GripStream. Run starts a goroutine
-// for each topic passed to the StreamWorker so that events for each topic are
-// processed in parallel. It will run until the context has been cancelled and
-// in the event of an error it will sleep and try again.
-func (s StreamWorker) Run(ctx context.Context) {
-	for _, topic := range s.topics {
-		s.log.Info("starting stream worker", "topic", topic)
-		go func(topic string) {
-			for {
-				select {
-				case <-ctx.Done():
-					s.log.Info("stopping stream worker", "err", ctx.Err())
-					return
-				default:
-					if err := s.processTopic(ctx, topic); err != nil {
-						if err == context.Canceled {
-							return
-						}
-						s.log.Error("stream worker failed", "topic", topic)
-						// We hit an error, sleep and try subscribing again
-						time.Sleep(10 * time.Second)
-					}
-
-					// If for some reason run exits without an error we just
-					// re-enter the loop and run again
-				}
-			}
-		}(topic)
+// Pub makes StreamWorker implement the golang sdks stream.EventStreamListener
+// interface.
+func (s StreamWorker) Pub(ctx context.Context, event stream.Event) error {
+	if event.SSEEvent == nil {
+		return errors.New("can't publish event with nil SSEEvent")
 	}
-}
-
-// processTopic performs the logic of listening for events on a Stream for
-// a topic and forwarding them on to the GripStream for that topic
-func (s StreamWorker) processTopic(ctx context.Context, topic string) error {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
-	events := s.subscribe(ctx, topic)
 
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case e, ok := <-events:
-			if !ok {
-				return nil
-			}
+	topic := event.Environment
+	content := fmt.Sprintf("event: *\ndata: %s\n\n", event.SSEEvent.Data)
 
-			if e.err != nil {
-				return e.err
-			}
-
-			if err := s.publish(ctx, e); err != nil {
-				return err
-			}
-			s.log.Debug("succesfully published event to pushpin", "topic", topic, "channel", e.channel, "content", e.content)
-		}
+	if err := s.publish(ctx, streamEvent{
+		channel: topic,
+		content: content,
+		err:     nil,
+	}); err != nil {
+		s.log.Error("stream worker failed", "topic", topic)
```

Comment:
Just wondering, should we be returning this error?

Reply:
For now I didn't and just log it here. If we return an error the go sdk will just log it again so wasn't sure it was helpful - https://github.com/harness/ff-golang-server-sdk/blob/main/stream/sse.go#L177